src/Umbraco.Web/Routing/RedirectTrackingEventHandler.cs

```diff
@@ -89,62 +90,92 @@ protected override void ApplicationStarted(UmbracoApplicationBase umbracoApplica
         {
-            get { return Moving && UmbracoContext.Current.HttpContext.Items[ContextKey2] != null; }
+            get
+            {
+                return Moving && RequestCache.GetCacheItem(ContextKey2) != null;
+            }
             set
             {
                 if (Moving && value)
-                    UmbracoContext.Current.HttpContext.Items[ContextKey2] = true;
+                {
+                    //this forces true into the cache
+                    RequestCache.GetCacheItem(ContextKey2, () => true);
+                }
                 else
-                    UmbracoContext.Current.HttpContext.Items.Remove(ContextKey2);
+                {
+                    RequestCache.ClearCacheItem(ContextKey2);
+                }
             }
         }
 
         private static bool Moving
         {
-            get { return UmbracoContext.Current.HttpContext.Items[ContextKey1] != null; }
+            get { return RequestCache.GetCacheItem(ContextKey1) != null; }
             set
             {
                 if (value)
-                    UmbracoContext.Current.HttpContext.Items[ContextKey1] = true;
+                {
+                    //this forces true into the cache
+                    RequestCache.GetCacheItem(ContextKey1, () => true);
+                }
                 else
                 {
-                    UmbracoContext.Current.HttpContext.Items.Remove(ContextKey1);
-                    UmbracoContext.Current.HttpContext.Items.Remove(ContextKey2);
+                    RequestCache.ClearCacheItem(ContextKey1);
+                    RequestCache.ClearCacheItem(ContextKey2);
                 }
             }
         }
 
+        /// <summary>
+        /// Before the items are published, we need to get it's current URL before it changes
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
         private static void ContentService_Publishing(IPublishingStrategy sender, PublishEventArgs<IContent> args)
         {
             if (LockedEvents) return;
 
-            var contentCache = UmbracoContext.Current.ContentCache;
+            var contentCache = GetPublishedCache();
+            if (contentCache == null) return;
+
             foreach (var entity in args.PublishedEntities)
             {
-                var entityContent = contentCache.GetById(entity.Id);
-                if (entityContent == null) continue;
-                foreach (var x in entityContent.DescendantsOrSelf())
+                //don't continue if this entity hasn't changed with regards to anything 
+                // that might change it's URLs
+                if (entity.IsDirty() == false) continue;
+
+                if (entity.IsPropertyDirty("Name") 
+                    || entity.IsPropertyDirty(Constants.Conventions.Content.UrlName)
+                    || entity.IsPropertyDirty(Constants.Conventions.Content.UrlAlias)
+                    || Moving)
```

Comment:
I understand what you are saying but I fail to see how we can check if we should or should not iterate over the entire descendants tree - this is something we simply cannot do. In the `Publishing` event, we can check what the item's currently UrlSegment is based on the `IUrlSegmentProvider` and then we can check if that has changed in `PageCacheRefresher_CacheUpdated` but there is no way to check if the segment has changed prior to `PageCacheRefresher_CacheUpdated` and therefore no way to determine if we need to iterate the entire descendant tree.

I'll try to do a horrible thing - store a delegate to iterate over the descendants of the `IPublishedContent` instance resolved during `ContentService_Publishing`, then check if the segment has changed in `PageCacheRefresher_CacheUpdated` and if so then execute the delegate stored from `ContentService_Publishing` and see if the `ContextualPublishedContentCache` stored in that delegate returns the old URLs. If so than we can make this work, otherwise I don't think this is possible without allowing the `IUrlSegmentProvider` to explicitly be able to return a value indicating if it has changed.